torch/csrc/jit/script/convert_to_ssa.cpp

```diff
@@ -265,21 +246,120 @@ struct SSATransformer {
+      Node* n = *it;
+      it++;
+      switch (n->kind()) {
+        case prim::If: {
+          assignExitContinuations(n->blocks().at(0));
+          assignExitContinuations(n->blocks().at(1));
+        } break;
+        case prim::Function: {
+          LoopExitContinuations closure_block;
+          closure_block.run(n->blocks().at(0));
+        } break;
+        case prim::Loop: {
+          Node* prev_exit = curr_loop_exit;
+          setCurrLoopExit(n->blocks().at(0));
+          assignExitContinuations(n->blocks().at(0));
+          curr_loop_exit = prev_exit;
+        } break;
+        case prim::ContinueStmt: {
+          auto loop_exit = graph->create(prim::LoopExit, 0)->insertAfter(n);
+          auto header_block = loop_exit->addBlock();
+          auto cur_loop = curr_loop_exit->owningBlock()->owningNode();
+          auto pre_header = cur_loop->blocks().at(1);
+          header_block->cloneFrom(pre_header, [](Value* v) { return v; });
+          moveBlockBeforeNode(n, header_block);
+          loop_exit->addInput(header_block->outputs().at(0));
+          loop_exit->eraseBlock(0);
+          addLoopCarriedOutputs(loop_exit);
+          n->destroy();
+        } break;
+        case prim::BreakStmt: {
+          auto loop_exit = graph->create(prim::LoopExit, 0)->insertAfter(n);
+          // first input is the loop continue condition - break sets false
+          loop_exit->addInput(false_val);
+          addLoopCarriedOutputs(loop_exit);
+          n->destroy();
+        } break;
+      }
+    }
+  }
+
+  void run(Block* b) {
+    {
+      graph = b->owningGraph();
+      WithInsertPoint guard(graph->block()->nodes().front());
+      false_val = graph->insertConstant(false);
+    }
+    assignExitContinuations(b);
+  }
+
+  void run(std::shared_ptr<Graph>& graph) {
+    run(graph->block());
+  }
+
+  Graph* graph;
+  Value* false_val;
+  Node* curr_loop_exit = nullptr;
+};
+
+// Converting to SSA works in multiple parts. First, we add control flow
+// loads and stores to the graph. Now that control flow outputs are set,
+// we can set remove Break & Continue to have the correct continuations to the
+// end of the block (LoopExit). Then we inline the loop condition into the
+// graph. Then, we erase Loads & Stores.
+// If & Loop block outputs have a prim::BlockExit node that designates the input
+//  values to be outputs of the block it is contained.
+// if cond:
+//    x = 1
+// ->
+// prim::If(%cond)
+//    x.1% = 1
+//    prim::BlockExit(%x.1)
+//
+// Finally, we unify LoopExit and BlockExit nodes to assign the correct
+// outputs to each block.
 void ConvertToSSA(std::shared_ptr<Graph>& graph) {
-  inlineLoopCondition(graph->block());
   ControlFlowLoadStores ctrl;
   ctrl.run(graph);
-  SSATransformer ssa;
-  ssa.run(graph);
+  LoopExitContinuations exit_vars;
+  exit_vars.run(graph);
+  InlineLoopCondition(graph);
+  EraseLoadsStores erase_loads_stores;
+  erase_loads_stores.run(graph);
+  TransformExits(graph);
```

Comment:
You probably have discussed this with @zdevito, but are we sure that we want to introduce `BlockExit` and `LoopExit` as a primary IR nodes? It looks like they are just used as a metadata to pass some info between these passes - could we just perform a book-keeping in some map and pass it instead? IR transformations to keep these nodes are responsibly for a bulk part of complexity in this PR in my opinion.

NB: `Load` and `Store` are in somewhat in similar position, but I think they are fine because 1) they have much clearer semantics, 2) they don't need such sophisticated IR manipulations.

Reply:
What book keeping would you like exactly? I think passing stateful maps around will be more complicated than something that is explicit in the IR & human readable from one pass to the next.

Reply:
The book-keeping I'm referring to is basically arguments of `BlockExit`/`LoopExit` that you encode in IR. The problem with the current approach is that even though it's supposed to be human readable, it really is not. I will have hard time understanding what a specific BlockExit means if I see it in IR. Does it depend on the context? What do arguments mean? What does it actually do (is it a jump?)?

Having stateful maps is generally bad, but if this is just between a couple of passes that are already dependent on each other and always run together, that might be not the worst.

Reply:
BlockExit jumps to the most recent block, LoopExit jumps to the most recent loop block. I'm happy to consider other options that would be more clear. Maybe other there other names you think would be better ?

Reply:
Can we just set the exit behavior as an attribute on the prim::Return node that already exists at the end of every block. It seems redundant to have both outputs of the block, and separate node temporarily the block to express those outputs. The attribute can just be a string specifying one of  `attr::exit_level = {"block", "loop", "function"}`

Reply:
ATM, there needs to be a distinction between the exits that are added in convert_to_ssa.cpp and the existing prim::Return for a couple reasons. 

1. We emit ifs with block outputs already added in compiler.cpp because of `and` and `or`. After convert_to_ssa and before exit_transforms is run it would look like.

```
  %z : bool = prim::If(%5)
    block0():
      %z.1 : bool = aten::__is__(%2, %20)
      = prim::Return[exit_level="block"]() # empty, added from convert_load_to_stores
      = prim::Return[exit_level="block"](%z.1) # added from compiler
    block1():
      %z.2 : bool = aten::__is__(%3, %4)
      = prim::Return[exit_level="block"]() # empty, added from convert_load_to_stores
      = prim::Return[exit_level="block"](%z.2) # added from compiler
```

 I'm not sure exactly how to handle this case because you have a block where two exits have a different number of returns, targeting the same block. This breaks invariants in exit_transform.


2. This is a smaller thing, but as seen above, this would break invariants invariants in ir.h that prim::Return does not have any nodes preceding it for LoopExits because will have nodes that are emitted after. You could special case LoopExits, maybe its a reason not to merge it with prim::Return. 


Neither of these are insurmountable it would just be added complexity to an already complex diff. IMO these changes would make this PR more complicated as a whole, (although you may disagree). I would prefer to unify things in a follow up.

Reply:
Why are already ssa form if statements getting transformed at all? Shouldn't the load/store additions just _add_ outputs to the existing if statement outputs? It also seems weird that there can be extra prim::Return nodes. There should only be a single prim return (`return_node()`) for each block. Nothing should be adding a prim::Return node. The second concern seems related to this point because we shouldn't be adding prim::Return nodes, just editing the existing prim::Return that serves as an output.

Reply:
My concern with "make the PR more complicated as a whole" is that it would still make the code complicated even if were across two PRs, and we should be looking for ways to handle all cases consistently so we do not incur this complexity. I think the root issue here is that already-ssa and needs to be converted to ssa if statements are not merged together in the ssa conversion pass.

Reply:
I don't think it's currently possible to unify the LoopExit to prim::Return node and to have one prim::Return per block because you need to have both a prim::Return for the block outputs and separately a Return for the LoopExit. 
E.g. 
```
if i == 1:
    break 
    a = 1
else:
    a = None
```
We unify `a = 1` and `a = None` to Optional[int]. Erasing the block output in the true branch would erase the original type captured, and set a to None in the resulting scope. 

Also, things get a little messy in the transform pass if block outputs can either exit to block or loop. In an if statement like above, where the true block exits to loop and the false block exits to block, when & how do you align the values that are escaping to loop scope and those that are escaping to block scope? 

You'd probably first transform LoopExits, then transform BlockExits, but then you're in state after transforming LoopExits where the BlockExits contain both the loop exit values and block exit values, without clearly distinguishing which is which.